vortex-serde/src/layouts/read/buffered.rs

```diff
@@ -1,105 +1,168 @@
 use std::collections::VecDeque;
+use std::mem;
 
+use croaring::Bitmap;
 use vortex::array::ChunkedArray;
 use vortex::compute::slice;
 use vortex::{Array, ArrayDType, IntoArray};
-use vortex_error::VortexResult;
+use vortex_error::{vortex_bail, VortexResult};
 
+use crate::layouts::read::selection::RowSelector;
 use crate::layouts::read::{LayoutReader, ReadResult};
+use crate::layouts::{Message, RangeResult};
+
+pub type RangedLayoutReader = ((usize, usize), Box<dyn LayoutReader>);
+pub type RangedArray = ((usize, usize), Array);
 
 #[derive(Debug)]
-pub struct BufferedReader {
-    layouts: VecDeque<Box<dyn LayoutReader>>,
-    arrays: VecDeque<Array>,
-    batch_size: usize,
+pub struct BufferedArrayReader {
+    layouts: VecDeque<RangedLayoutReader>,
+    arrays: VecDeque<RangedArray>,
+    next_range_offset: usize,
 }
 
-impl BufferedReader {
-    pub fn new(layouts: VecDeque<Box<dyn LayoutReader>>, batch_size: usize) -> Self {
+impl BufferedArrayReader {
+    pub fn new(layouts: VecDeque<RangedLayoutReader>) -> Self {
         Self {
             layouts,
-            arrays: Default::default(),
-            batch_size,
+            arrays: VecDeque::new(),
+            next_range_offset: 0,
         }
     }
 
     fn is_empty(&self) -> bool {
         self.layouts.is_empty() && self.arrays.is_empty()
     }
 
-    fn buffered_row_count(&self) -> usize {
-        self.arrays.iter().map(Array::len).sum()
-    }
+    pub fn next_range(&mut self) -> VortexResult<RangeResult> {
+        if self.next_range_offset == self.layouts.len() {
+            return Ok(RangeResult::Rows(None));
+        }
 
-    fn buffer(&mut self) -> VortexResult<Option<ReadResult>> {
-        while self.buffered_row_count() < self.batch_size {
-            if let Some(mut layout) = self.layouts.pop_front() {
-                if let Some(rr) = layout.read_next()? {
-                    self.layouts.push_front(layout);
-                    match rr {
-                        read_more @ ReadResult::ReadMore(..) => {
-                            return Ok(Some(read_more));
-                        }
-                        ReadResult::Batch(a) => self.arrays.push_back(a),
+        match self.layouts[self.next_range_offset].1.next_range()? {
+            RangeResult::ReadMore(m) => Ok(RangeResult::ReadMore(m)),
+            RangeResult::Rows(r) => match r {
+                None => {
+                    self.next_range_offset += 1;
+                    self.next_range()
+                }
+                Some(rs) => {
+                    let layout_range = self.layouts[self.next_range_offset].0;
```

Comment:
I think it would help legibility to bind `self.layouts[self.next_range_offset]` above. Is this something like `next_layout`?

Reply:
It's a bit annoying since you can't without a clone, that's why this is inlined since .0 is a copy value

Reply:
I added a comment